test/Microsoft.AspNetCore.SignalR.Redis.Tests/RedisEndToEnd.cs

```diff
@@ -0,0 +1,129 @@
+            }
+        }
+
+        [ConditionalTheory]
+        [SkipIfDockerNotPresent]
+        [MemberData(nameof(TransportTypesAndProtocolTypes))]
+        public async Task HubConnectionCanSendAndReceiveGroupMessages(TransportType transportType, Type protocolType)
+        {
+            using (StartLog(out var loggerFactory, testName:
+                $"{nameof(HubConnectionCanSendAndReceiveGroupMessages)}_{transportType.ToString()}_{protocolType.Name}"))
+            {
+                var connection = CreateConnection(_serverFixture.FirstServer.Url + "/echo", transportType, protocolType, loggerFactory);
+                var secondConnection = CreateConnection(_serverFixture.SecondServer.Url + "/echo", transportType, protocolType, loggerFactory);
+
+                var tcs = new TaskCompletionSource<string>();
+                connection.On<string>("Echo", message => tcs.TrySetResult(message));
+                var tcs2 = new TaskCompletionSource<string>();
+                secondConnection.On<string>("Echo", message => tcs2.TrySetResult(message));
+
+                await secondConnection.StartAsync().OrTimeout();
+                await connection.StartAsync().OrTimeout();
+                await connection.InvokeAsync("EchoGroup", "Test", "Hello world").OrTimeout();
+
+                Assert.Equal("Hello world", await tcs.Task.OrTimeout());
+                Assert.Equal("Hello world", await tcs2.Task.OrTimeout());
+
+                await connection.DisposeAsync().OrTimeout();
+            }
+        }
+
+        private static HubConnection CreateConnection(string url, TransportType transportType, Type protocolType, ILoggerFactory loggerFactory)
+        {
+            var builder = new HubConnectionBuilder()
+                .WithUrl(url)
+                .WithTransport(transportType)
+                .WithLoggerFactory(loggerFactory);
+            if (protocolType == typeof(MessagePackHubProtocol))
+            {
+                builder.WithMessagePackProtocol();
+            }
+            else if (protocolType == typeof(JsonHubProtocol))
+            {
+                builder.WithJsonProtocol();
+            }
+
+            return builder.Build();
+        }
+
+        public static IEnumerable<object[]> TransportTypes
+        {
+            get
+            {
+                if (TestHelpers.IsWebSocketsSupported())
+                {
+                    yield return new object[] { TransportType.WebSockets };
+                }
+                yield return new object[] { TransportType.ServerSentEvents };
+                yield return new object[] { TransportType.LongPolling };
+            }
+        }
+
+        public static IEnumerable<object[]> TransportTypesAndProtocolTypes
+        {
+            get
+            {
+                foreach (var transport in TransportTypes)
+                {
+                    yield return new object[] { transport[0], typeof(JsonHubProtocol) };
```

Comment:
create Protocol instances instead of passing types to be able to use just `.WithProtocol()` when creating the client.